File: Lecture 05/Python Code/OCP/OCPFollowed.py
from abc import ABC, abstractmethod
class Product:
    def __init__(self, name, price):
        self.name = name
        self.price = price

class ShoppingCart:
    def __init__(self):
        self.products = []

    def add_product(self, product):
        self.products.append(product)

    def calculate_total(self):
        return sum(p.price for p in self.products)

class ShoppingCartPrinter:
    def __init__(self, cart):
        self.cart = cart

    def print_invoice(self):
        print("Shopping Cart Invoice:")
        for product in self.cart.products:
            print(f"{product.name} - Rs {product.price}")
        print(f"Total: Rs {self.cart.calculate_total()}")

class Persistence(ABC):
    @abstractmethod
    def save(self, cart):
        pass

class SQLPersistence(Persistence):
    def save(self, cart):
        print("Saving shopping cart to SQL DB...")

class MongoPersistence(Persistence):
    def save(self, cart):
        print("Saving shopping cart to MongoDB...")

class FilePersistence(Persistence):
    def save(self, cart):
        print("Saving shopping cart to a file...")

if __name__ == "__main__":
    cart = ShoppingCart()
    cart.add_product(Product("Laptop", 50000))
    cart.add_product(Product("Mouse", 2000))

    printer = ShoppingCartPrinter(cart)
    printer.print_invoice()

    sql_persistence = SQLPersistence()
    mongo_persistence = MongoPersistence()
    file_persistence = FilePersistence()

    sql_persistence.save(cart)
    mongo_persistence.save(cart)
    file_persistence.save(cart)

File: Lecture 05/Typescript/OCP/OCPFollowed.ts
// Product class representing any item in eCommerce.
class Product {
    public name: string
    public price: number

    constructor(name: string, price: number) {
        this.name = name
        this.price = price
    }
}

// 1. ShoppingCart: Only responsible for cart related business logic.
class ShoppingCart {
    private products: Product[] = []

    public addProduct(p: Product): void {
        this.products.push(p)
    }

    public getProducts(): Product[] {
        return this.products
    }

    // Calculates total price in cart.
    public calculateTotal(): number {
        let total = 0
        for (const p of this.products) {
            total += p.price
        }
        return total
    }
}

// 2. ShoppingCartPrinter: Only responsible for printing invoices
class ShoppingCartPrinter {
    private cart: ShoppingCart

    constructor(cart: ShoppingCart) {
        this.cart = cart
    }

    public printInvoice(): void {
        console.log("Shopping Cart Invoice:")
        for (const p of this.cart.getProducts()) {
            console.log(`${p.name} - Rs ${p.price}`)
        }
        console.log(`Total: Rs ${this.cart.calculateTotal()}`)
    }
}

// Persistence abstraction
interface Persistence {
    save(cart: ShoppingCart): void
}

class SQLPersistence implements Persistence {
    save(cart: ShoppingCart): void {
        console.log("Saving shopping cart to SQL DB...")
    }
}

class MongoPersistence implements Persistence {
    save(cart: ShoppingCart): void {
        console.log("Saving shopping cart to MongoDB...")
    }
}

class FilePersistence implements Persistence {
    save(cart: ShoppingCart): void {
        console.log("Saving shopping cart to a file...")
    }
}

// Equivalent of Java main
function main(): void {
    const cart = new ShoppingCart()
    cart.addProduct(new Product("Laptop", 50000))
    cart.addProduct(new Product("Mouse", 2000))

    const printer = new ShoppingCartPrinter(cart)
    printer.printInvoice()

    const db: Persistence = new SQLPersistence()
    const mongo: Persistence = new MongoPersistence()
    const file: Persistence = new FilePersistence()

    db.save(cart)
    mongo.save(cart)
    file.save(cart)
}

main()


/**
 * If a .ts file has no imports or exports, TypeScript treats it as a global script.
 * Adding `export {}` marks the file as a module and prevents
 * "Duplicate identifier 'ShoppingCart'" errors.
 */
export {}